From c00edd07b91d1abfc29563e58e89789ebee4baa5 Mon Sep 17 00:00:00 2001
From: Wen Gong <wgong@codeaurora.org>
Date: Fri, 26 Mar 2021 18:43:14 +0800
Subject: [PATCH 09/16] mac80211: add more type for prevent mixed key and
 fragment cache attacks

For some chips, ieee80211_has_protected() is not meet, and it also need
do protection for cache attack. This fixs CVE-2020-24587 for the chips.

Tested-on: QCA6174 hw3.2 PCI WLAN.RM.4.4.1-00110-QCARMSWP-1

Change-Id: I4c1cb7e21417f57ba9cd16bbd317abdefd1432d3
Signed-off-by: Wen Gong <wgong@codeaurora.org>
---
 net/mac80211/rx.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/net/mac80211/rx.c b/net/mac80211/rx.c
index 4c28fe6..5753760 100644
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -2417,6 +2417,8 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 				     sizeof(rx->key->u.gcmp.rx_pn[queue]));
 			BUILD_BUG_ON(IEEE80211_CCMP_PN_LEN !=
 				     IEEE80211_GCMP_PN_LEN);
+		} else if (rx->key) {
+			entry->key_color = rx->key->color;
 		}
 		return RX_QUEUED;
 	}
@@ -2454,6 +2456,8 @@ ieee80211_rx_h_defragment(struct ieee80211_rx_data *rx)
 		if (memcmp(pn, rpn, IEEE80211_CCMP_PN_LEN))
 			return RX_DROP_UNUSABLE;
 		memcpy(entry->last_pn, pn, IEEE80211_CCMP_PN_LEN);
+	} else if (rx->key && entry->key_color != rx->key->color) {
+		return RX_DROP_UNUSABLE;
 	}
 
 	skb_pull(rx->skb, ieee80211_hdrlen(fc));
-- 
2.7.4

